### ISSUE TITLE
False positive `invalid-name` when using assignment expression inside of a comprehension

### ISSUE THREAD

**Issue by wshanks** [PyCQA/pylint#5242]
### Bug description

```python
"""docstring"""


def func():
    "docstring"
    nonzero = [
        o
        for t in [0, 1, 2, 0, 3]
        if (o := t) != 0
    ]
    print(nonzero)
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
************* Module test
test.py:9:12: C0103: Variable name "o" doesn't conform to snake_case naming style (invalid-name)
```

### Expected behavior

No warnings or error for this code. Note: pylint 2.10 produces no warnings or error. The warning only appears for 2.11.

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.4
Python 3.9.7 | packaged by conda-forge | (default, Sep  2 2021, 17:58:34) 
[GCC 9.4.0]
```

### OS / Environment

Fedora 34, using conda-forge

**Comment by timmartin** [comment]
It looks to me like this is working correctly. The standard snake_case requirement for variables is to be at least 3 characters long. If you change your code to:

    """docstring"""
    
    
    def func():
        "docstring"
        nonzero = [
            ooo
            for t in [0, 1, 2, 0, 3]
            if (ooo := t) != 0
        ]
        print(nonzero)

then it doesn't trigger the error. One source of confusion could be that the variable name `t` is an `inlinevar` rather than a `variable`, which has looser rules (i.e. no restriction by default). This makes sense to me because `t` is scoped to just the comprehension, while `o` is in scope to the end of the function.

**Comment by Pierre-Sassoulas** [comment]
Variable inside comprehension are supposed to be allowed to be one letter variable, I guess the walrus is messing with that.

**Comment by timmartin** [comment]
It makes sense to me that variables scoped entirely within the comprehension can have shorter names, but that's not what's happening here. The walrus operator is creating a variable that lives on beyond the comprehension and pollutes the namespace for the rest of the function. I don't see why you'd want the rules here to be any different than if you used the walrus operator to assign a variable without a comprehension.

**Comment by Pierre-Sassoulas** [comment]
Ho my bad, you're right ! Closing as it's not a bug.

**Comment by wshanks** [comment]
Thanks for the response @timmartin! This behavior goes against my intuition but pylint is rightly interpreting the code here. I looked at PEP 572 again and found that this exact case [is discussed](https://www.python.org/dev/peps/pep-0572/#why-not-use-a-sublocal-scope-and-prevent-namespace-pollution).

It makes me ponder about what best practice should be for assignment expressions within comprehensions. I don't think it is a good idea to define a variable within a comprehension that you want to use afterwards. I could imagine a pylint rule that allows single letter assignment expression variables within comprehensions as long as they are not referenced outside of the comprehensions. Or maybe I will just go back to for loops :slightly_smiling_face:

**Comment by timmartin** [comment]
Based on the way the code works at the moment, I think it would be quite hard to have an `invalid-name` warning that only triggers if the variable is used later. If someone is sufficiently motivated I'm sure they could do it, though.

More realistically, I think I'd favour it being a non-default but configurable rule that would warn if a variable is defined inside a comprehension this way and then used outside. It can then be a question for each individual project to choose their coding standard. We could combine this with a different class of naming rule for variables defined inline within a comprehension, so projects could either:

- disallow variables defined within the comprehension from being used elsewhere, and reconfigure the naming rules to allow single-character names on this class of variable
- allow variables defined this way to be used outside the comprehension, and leave the naming rules at the default (same as any other variable).

This pushes more work onto the person configuring the Pylint rules, of course.
